### PR TITLE
Request response browser2

### PR DESCRIPTION
**Pull request by bretambrose**
* Subscription manager implementation and tests.  Both impl and tests are direct ports from the corresponding pieces of aws-c-mqtt

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lib/browser/mqtt_request_response/protocol_adapter.ts [hunk 1]
```diff
@@ -461,4 +469,17 @@ export class ProtocolClientAdapter extends BufferedEventEmitter {
                 return false;
         }
     }
+
+    private static isSubackReasonCodeRetryable(reasonCode: mqtt5.SubackReasonCode) : boolean {
+        switch (reasonCode) {
+            case mqtt5.SubackReasonCode.UnspecifiedError:
```
Comment (sbSteveK) [inline lib/browser/mqtt_request_response/protocol_adapter.ts:475]: Debatable: UnspecifiedError seems like something we don't want to retry. Seems like the kind of thing where a retry wouldn't resolve but I don't have strong enough feelings about it if you put it here for a reason.

File: lib/browser/mqtt_request_response/subscription_manager.ts [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+        this.adapter.addListener(protocol_adapter.ProtocolClientAdapter.SUBSCRIBE_COMPLETION, this.handleSubscribeCompletionEvent.bind(this));
+        this.adapter.addListener(protocol_adapter.ProtocolClientAdapter.UNSUBSCRIBE_COMPLETION, this.handleUnsubscribeCompletionEvent.bind(this));
+        this.adapter.addListener(protocol_adapter.ProtocolClientAdapter.CONNECTION_STATUS, this.handleConnectionStatusEvent.bind(this));
+    }
+
+    close() {
+        this.closed = true;
+
+        this.unsubscribeAll();
+    }
```
Comment (sfod) [inline lib/browser/mqtt_request_response/subscription_manager.ts:191]: Not sure if this is an issue or not, but I think, there is a possibility for a subscription to succeed after the `close` is called. Then, that subscription will remain active after the manager is closed.

Here is a hypothetical case:
1. User calls `manager.acquireSubscription("a/b")` --> `this.activateSubscription("a/b")` --> `this.adapter.subscribe("a/b")`
2. User calls `manager.close()` --> `this.unsubscribe("a/b")` --> `this.adapter.unsubscribe("a/b")`
3. MQTT broker processes `SUBSCRIBE` and returns a retryable SUBACK --> `this.activateSubscription("a/b")` --> `this.adapter.subscribe("a/b")`
4. MQTT broker processes `UNSUBSCRIBE`.
5. MQTT broker processes `SUBSCRIBE` with a success.

Even without the `retryable` logic, it's possible that MQTT broker will process subscribe and unsubscribe requests in reversed order:
3. MQTT broker processes `UNSUBSCRIBE`
4. MQTT broker processes `SUBSCRIBE`
But I don't think we can handle this in a robust way.

But the retryable logic probably could be improved by checking `closed` flag.

Comment (bretambrose) [reply lib/browser/mqtt_request_response/subscription_manager.ts:191]: Predicating the resubscribe with a check of this.closed would address, yes?

Beyond that I am unsure how the broker would receive an unsubscribe and a subscribe out of submission order.

Comment (sfod) [reply lib/browser/mqtt_request_response/subscription_manager.ts:191]: > Predicating the resubscribe with a check of this.closed would address, yes?

I believe yes, it should help.

File: lib/browser/mqtt_request_response/protocol_adapter.ts (hunk 1 above)
Comment (bretambrose) [reply lib/browser/mqtt_request_response/protocol_adapter.ts:475]: Removed

File: lib/browser/mqtt_request_response/protocol_adapter.ts [hunk 3]
```diff
@@ -461,4 +469,17 @@ export class ProtocolClientAdapter extends BufferedEventEmitter {
                 return false;
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply lib/browser/mqtt_request_response/protocol_adapter.ts:469]: Seems like a good idea.

**Pull request closed without merging**